### PR TITLE
Various last PY3 fixes

### PR DESCRIPTION
**Pull request by gpotter2**
- fix another tls_version failing usage on python 3(actually all the ones used in `dispatch_hook` functions should have the not None test)
- cleanup mock_windows.uts
- fix IPython main.interact() test + increase coverage

### CODE REVIEW

**Comment by gpotter2** [comment]
@guedou @p-l- With this PR, tests should be passing on python 3

**Comment by guedou** [comment]
I don’t understand why this new (fake ?) TLS header is needed. Could you give us explanations ? For example, it seems that Raw could also work.

**Comment by gpotter2** [comment]
@guedou You're actually very right, this is completly unneeded. Switched to Raw

File: scapy/layers/tls/handshake.py [hunk 1]
```diff
@@ -1179,7 +1179,7 @@ class TLSNewSessionTicket(_TLSHandshake):
     @classmethod
     def dispatch_hook(cls, _pkt=None, *args, **kargs):
         s = kargs.get("tls_session", None)
-        if s and s.tls_version >= 0x0304:
+        if s and s.tls_version and s.tls_version >= 0x0304:
```
Comment (guedou) [inline scapy/layers/tls/handshake.py]: Did you find a bug with this line ? It will be nice to add a test.
I guess that it would be better to test that s is an instance of `tlsSession`

File: scapy/layers/tls/record.py [hunk 2]
```diff
@@ -288,6 +287,8 @@ def dispatch_hook(cls, _pkt=None, *args, **kargs):
                     if s.rcs and not isinstance(s.rcs.cipher, Cipher_NULL):
                         from scapy.layers.tls.record_tls13 import TLS13
                         return TLS13
+        if _pkt and len(_pkt) < 5: # TLS Continuation packet
```
Comment (guedou) [inline scapy/layers/tls/record.py]: Please use two spaces before `:` and `#`.

Comment (guedou) [reply scapy/layers/tls/record.py]: Also, I wonder if this comment is useful. It might be better to remove it.

File: test/regression.uts [hunk 3]
```diff
@@ -213,25 +213,30 @@ assert(len(conf.temp_files) == 0)
 = Emulate interact()
 
 import mock, sys
+from scapy.main import interact
```
Comment (guedou) [inline test/regression.uts]: Awesome fix. I did not manage to fix it myself =)

File: scapy/layers/tls/handshake.py (hunk 1 above)
Comment (gpotter2) [reply scapy/layers/tls/handshake.py]: Test added !

**Comment by gpotter2** [comment]
Please restart appveyor (could not start)

File: scapy/layers/tls/handshake.py (hunk 1 above)
Comment (guedou) [reply scapy/layers/tls/handshake.py]: Can `s.tls_version` be None ? If the answer is yes, could you add a test to check for this specific corner case ?

Comment (gpotter2) [reply scapy/layers/tls/handshake.py]: Already included in https://github.com/secdev/scapy/pull/992/files#diff-cd90a85f0ec6440949616e4896487acaR14

Comment (guedou) [reply scapy/layers/tls/handshake.py]: Sorry, I meant rewrite line 1182 as:
`if s and isinstance(s, tlsSession) and s.tls_version >= 0x0304:`
and make sure that the unit test sets the "tls_session" kwargs as None to check that this new test works.

Comment (gpotter2) [reply scapy/layers/tls/handshake.py]: Ok i've added an explicit test :)

File: scapy/layers/tls/record.py [hunk 4]
```diff
@@ -288,6 +287,10 @@ def dispatch_hook(cls, _pkt=None, *args, **kargs):
                     if s.rcs and not isinstance(s.rcs.cipher, Cipher_NULL):
                         from scapy.layers.tls.record_tls13 import TLS13
                         return TLS13
+        if _pkt and len(_pkt) < 5:
+            if _pkt != b"\x00":  # record is often b'\x00', used as a separator
```
Comment (guedou) [inline scapy/layers/tls/record.py]: Is this test needed ? I do not understand you comment.

Comment (gpotter2) [reply scapy/layers/tls/record.py]: I have found no doc saying so, but from what i've sniffed, it's very common to get some of those 
 kinda empty packets on a network (for example mine), which are used as padding, and we don't want the warning message to be triggered.

Comment (guedou) [reply scapy/layers/tls/record.py]: I am really curious. Could you provide an example ?

Comment (gpotter2) [reply scapy/layers/tls/record.py]: This might not be a good fix. For instance, here's a pcap file obtained with:
`wrpcap("test.pcap", sniff(lfilter=lambda x: TCP in x and 443 in [x.dport, x.sport]))`

[test.pcap.zip](https://github.com/secdev/scapy/files/1583047/test.pcap.zip)

```
>>> load_layer("tls")
>>> conf.debug_dissector = 1
>>> rdpcap("test.pcap")
>>> rdpcap("test.pcap")
Traceback (most recent call last):
  File "scapy\layers\tls\record.py", line 99, in m2i
    return cls(m, tls_session=pkt.tls_session)
  File "scapy\base_classes.py", line 231, in __call__
    i.__init__(*args, **kargs)
  File "scapy\layers\tls\session.py", line 833, in __init__
    **fields)
  File "scapy\packet.py", line 127, in __init__
    self.dissect(_pkt)
  File "scapy\packet.py", line 734, in dissect
    s = self.do_dissect(s)
  File "scapy\packet.py", line 697, in do_dissect
    s, fval = f.getfield(self, s)
  File "scapy\layers\tls\extensions.py", line 626, in getfield
    return s[l:], self.m2i(pkt, s[:l])
  File "scapy\layers\tls\extensions.py", line 647, in m2i
    l = struct.unpack("!H", m[2:4])[0]
error: unpack requires a string argument of length 2
Traceback (most recent call last):
  File "scapy\layers\tls\record.py", line 99, in m2i
    return cls(m, tls_session=pkt.tls_session)
  File "scapy\base_classes.py", line 231, in __call__
    i.__init__(*args, **kargs)
  File "scapy\layers\tls\session.py", line 833, in __init__
    **fields)
  File "scapy\packet.py", line 127, in __init__
    self.dissect(_pkt)
  File "scapy\packet.py", line 734, in dissect
    s = self.do_dissect(s)
  File "scapy\packet.py", line 697, in do_dissect
    s, fval = f.getfield(self, s)
  File "scapy\fields.py", line 421, in getfield
    i = self.m2i(pkt, s)
  File "scapy\layers\tls\keyexchange.py", line 262, in m2i
    p = cls(m, tls_session=s)
TypeError: 'NoneType' object is not callable
ERROR: TLS dissector failed
ERROR: TCP dissector failed
ERROR: more IP dissector failed
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-4-cbc8d10a730b> in <module>()
----> 1 rdpcap("test.pcap")

Z:\Coding\github\scapy\scapy\utils.py in rdpcap(filename, count)
    705     """
    706     with PcapReader(filename) as fdesc:
--> 707         return fdesc.read_all(count=count)
    708
    709

Z:\Coding\github\scapy\scapy\utils.py in read_all(self, count)
    884         return p
    885     def read_all(self,count=-1):
--> 886         res = RawPcapReader.read_all(self, count)
    887         from scapy import plist
    888         return plist.PacketList(res,name = os.path.basename(self.filename))

Z:\Coding\github\scapy\scapy\utils.py in read_all(self, count)
    835         while count != 0:
    836             count -= 1
--> 837             p = self.read_packet()
    838             if p is None:
    839                 break

Z:\Coding\github\scapy\scapy\utils.py in read_packet(self, size)
    874
    875         try:
--> 876             p = self.LLcls(s)
    877         except KeyboardInterrupt:
    878             raise

Z:\Coding\github\scapy\scapy\base_classes.pyc in __call__(cls, *args, **kargs)
    229                 cls = config.conf.raw_layer
    230         i = cls.__new__(cls, cls.__name__, cls.__bases__, cls.__dict__)
--> 231         i.__init__(*args, **kargs)
    232         return i
    233

Z:\Coding\github\scapy\scapy\packet.pyc in __init__(self, _pkt, post_transform, _internal, _underlayer, **fields)
    125         self.raw_packet_cache_fields = None
    126         if _pkt:
--> 127             self.dissect(_pkt)
    128             if not _internal:
    129                 self.dissection_done(self)

Z:\Coding\github\scapy\scapy\packet.pyc in dissect(self, s)
    737
    738         payl,pad = self.extract_padding(s)
--> 739         self.do_dissect_payload(payl)
    740         if pad and conf.padding:
    741             self.add_payload(conf.padding_layer(pad))

Z:\Coding\github\scapy\scapy\packet.pyc in do_dissect_payload(self, s)
    715             cls = self.guess_payload_class(s)
    716             try:
--> 717                 p = cls(s, _internal=1, _underlayer=self)
    718             except KeyboardInterrupt:
    719                 raise

Z:\Coding\github\scapy\scapy\base_classes.pyc in __call__(cls, *args, **kargs)
    229                 cls = config.conf.raw_layer
    230         i = cls.__new__(cls, cls.__name__, cls.__bases__, cls.__dict__)
--> 231         i.__init__(*args, **kargs)
    232         return i
    233

Z:\Coding\github\scapy\scapy\packet.pyc in __init__(self, _pkt, post_transform, _internal, _underlayer, **fields)
    125         self.raw_packet_cache_fields = None
    126         if _pkt:
--> 127             self.dissect(_pkt)
    128             if not _internal:
    129                 self.dissection_done(self)

Z:\Coding\github\scapy\scapy\packet.pyc in dissect(self, s)
    737
    738         payl,pad = self.extract_padding(s)
--> 739         self.do_dissect_payload(payl)
    740         if pad and conf.padding:
    741             self.add_payload(conf.padding_layer(pad))

Z:\Coding\github\scapy\scapy\packet.pyc in do_dissect_payload(self, s)
    715             cls = self.guess_payload_class(s)
    716             try:
--> 717                 p = cls(s, _internal=1, _underlayer=self)
    718             except KeyboardInterrupt:
    719                 raise

Z:\Coding\github\scapy\scapy\base_classes.pyc in __call__(cls, *args, **kargs)
    229                 cls = config.conf.raw_layer
    230         i = cls.__new__(cls, cls.__name__, cls.__bases__, cls.__dict__)
--> 231         i.__init__(*args, **kargs)
    232         return i
    233

Z:\Coding\github\scapy\scapy\packet.pyc in __init__(self, _pkt, post_transform, _internal, _underlayer, **fields)
    125         self.raw_packet_cache_fields = None
    126         if _pkt:
--> 127             self.dissect(_pkt)
    128             if not _internal:
    129                 self.dissection_done(self)

Z:\Coding\github\scapy\scapy\packet.pyc in dissect(self, s)
    737
    738         payl,pad = self.extract_padding(s)
--> 739         self.do_dissect_payload(payl)
    740         if pad and conf.padding:
    741             self.add_payload(conf.padding_layer(pad))

Z:\Coding\github\scapy\scapy\packet.pyc in do_dissect_payload(self, s)
    715             cls = self.guess_payload_class(s)
    716             try:
--> 717                 p = cls(s, _internal=1, _underlayer=self)
    718             except KeyboardInterrupt:
    719                 raise

Z:\Coding\github\scapy\scapy\base_classes.pyc in __call__(cls, *args, **kargs)
    229                 cls = config.conf.raw_layer
    230         i = cls.__new__(cls, cls.__name__, cls.__bases__, cls.__dict__)
--> 231         i.__init__(*args, **kargs)
    232         return i
    233

Z:\Coding\github\scapy\scapy\layers\tls\record.py in __init__(self, *args, **kargs)
    266     def __init__(self, *args, **kargs):
    267         self.deciphered_len = kargs.get("deciphered_len", None)
--> 268         super(TLS, self).__init__(*args, **kargs)
    269
    270     @classmethod

Z:\Coding\github\scapy\scapy\layers\tls\session.py in __init__(self, _pkt, post_transform, _internal, _underlayer, tls_session, **fields)
    831         Packet.__init__(self, _pkt=_pkt, post_transform=post_transform,
    832                         _internal=_internal, _underlayer=_underlayer,
--> 833                         **fields)
    834
    835     def __getattr__(self, attr):

Z:\Coding\github\scapy\scapy\packet.pyc in __init__(self, _pkt, post_transform, _internal, _underlayer, **fields)
    125         self.raw_packet_cache_fields = None
    126         if _pkt:
--> 127             self.dissect(_pkt)
    128             if not _internal:
    129                 self.dissection_done(self)

Z:\Coding\github\scapy\scapy\packet.pyc in dissect(self, s)
    730
    731     def dissect(self, s):
--> 732         s = self.pre_dissect(s)
    733
    734         s = self.do_dissect(s)

Z:\Coding\github\scapy\scapy\layers\tls\record.py in pre_dissect(self, s)
    377         """
    378         if len(s) < 5:
--> 379             raise Exception("Invalid record: header is too short.")
    380
    381         msglen = struct.unpack('!H', s[3:5])[0]

Exception: Invalid record: header is too short.
```

Comment (guedou) [reply scapy/layers/tls/record.py]: It works fine with 2.3.3.dev957 with Python 2 and Python 3 without cryptography installed:
```
In [1]: load_layer("tls"); conf.debbug_dissector=1 ; rdpcap("test_tls.pcap")
INFO: Can't import python-cryptography v1.7+. Disabled PKI & TLS crypto-related features.
WARNING: TLS ciphersuite not useable. Is the cryptography Python module installed ?
WARNING: TLS ciphersuite not useable. Is the cryptography Python module installed ?
WARNING: more TLS ciphersuite not useable. Is the cryptography Python module installed ?
WARNING: Unknown cipher suite 13053006 from ClientMasterKey
WARNING: Unknown cipher suite 1684248 from ClientMasterKey
WARNING: more Unknown cipher suite 3924658 from ClientMasterKey
Out[1]: <test_tls.pcap: TCP:1817 UDP:0 ICMP:0 Other:0>
```
I tested with Python2 and cryptography and it works fine.

Comment (gpotter2) [reply scapy/layers/tls/record.py]: If so my fix is wrong. Thanks !
Will investigate

Comment (gpotter2) [reply scapy/layers/tls/record.py]: Will do a separate PR

**Comment by gpotter2** [comment]
@p-l- If you are performing PY3 tests, you might want to use this patch for Windows. If fixes small things

File: scapy/layers/tls/handshake.py (hunk 1 above)
Comment (gpotter2) [reply scapy/layers/tls/handshake.py]: Sorry I had not understood what you meant, I added the `isinstance` test

File: test/regression.uts [hunk 5]
```diff
@@ -791,9 +791,11 @@ assert(len(tmp) == 1)
 = Sending an ICMP message 'forever' at layer 2 and layer 3
 ~ netaccess IP ICMP
 tmp = srloop(IP(dst="8.8.8.8")/ICMP(), count=1)
+tmp
```
Comment (p-l-) [inline test/regression.uts]: how is this helping? what is it supposed to trigger?

Comment (gpotter2) [reply test/regression.uts]: Not really helpful. Removed

**Comment by guedou** [comment]
Can we close this PR?

**Comment by gpotter2** [comment]
I've extracted the IPython fix to a new pr (#1031)
